src/Compilers/CSharp/Test/Emit2/Semantics/PatternMatchingTests.cs

```diff
@@ -11784,5 +11784,57 @@ public void Test2(G<object> x2)
                 Diagnostic(ErrorCode.ERR_BadBinaryOps, "x2 as T").WithArguments("as", "G<object>", "T").WithLocation(11, 18)
                 );
         }
+
+        [Fact]
+        [WorkItem(63476, "https://github.com/dotnet/roslyn/issues/63476")]
+        public void PatternNonConstant_UserDefined_ConvertionToInputType()
+        {
+            var source =
+@"
+class A {
+    public string S { get; set; }
+    public static implicit operator A(string s) { return new A { S = s }; }
+}
+class C
+{
+    static bool M(A a) => a switch { ""implicitA"" => true, _ => false };
```

Comment:
Hey good catch, I have extended the code to test for ExplicitUserDefined conversions too since its the user defined part that is problematic and generalised the error message to make sense for both implicit and explicit.